File: TsLan/src/lexer/lexer.ts
import { TokenItem, TokenType, Token, readIDENT, identNumber } from "../token/token";



const a = 'a'.charCodeAt(0)
const A = 'A'.charCodeAt(0)
const z = 'z'.charCodeAt(0)
const Z = 'Z'.charCodeAt(0)
const _ = '_'.charCodeAt(0)
const _0 = '0'.charCodeAt(0)
const _9 = '9'.charCodeAt(0)
const dot = '.'.charCodeAt(0)
export class Lexer {
    private input: string
    private pos: number
    private readPos: number
    private char: string

    constructor(input: string) {
        this.input = input
        this.pos = 0
        this.readPos = 0
        this.char = input[this.readPos]
        this.readChar()
    }


    private readChar() {
        if (this.readPos >= this.input.length) {
            this.char = '\0'
        } else {
            this.char = this.input[this.readPos]
        }

        this.pos = this.readPos
        this.readPos++
    }


    public nextToken(): Token {
        var tok: Token

        this.skipWhiteSpace();
        switch (this.char) {
            case '=':
                if (this.peekChar() == "=") {
                    tok = this.newToken(TokenType.EQUAL, this.input[this.pos] + this.input[this.readPos])
                } else {
                    tok = this.newToken(TokenType.ASSIGN, this.char)
                }
                break
            case ';':
                tok = this.newToken(TokenType.SEMICOLON, this.char)
                break
            case '(':
                tok = this.newToken(TokenType.LPAREN, this.char)
                break;
            case ')':
                tok = this.newToken(TokenType.RPAREN, this.char)
                break
            case ',':
                tok = this.newToken(TokenType.COMMA, this.char)
                break
            case '+':
                tok = this.newToken(TokenType.PLUS, this.char)
                break
            case '{':
                tok = this.newToken(TokenType.LBRACE, this.char)
                break
            case '}':
                tok = this.newToken(TokenType.RBRACE, this.char)
                break
            case '*':
                tok = this.newToken(TokenType.ASTERISK, this.char)
                break
            case '-':
                tok = this.newToken(TokenType.MINUS, this.char)
                break
            case '<':
                if (this.peekChar() == "=") {
                    tok = this.newToken(TokenType.LTEQUAL, this.input[this.pos] + this.input[this.readPos])
                } else {
                    tok = this.newToken(TokenType.LT, this.char)
                }
                break
            case '>':
                if (this.peekChar() == "=") {
                    tok = this.newToken(TokenType.GTEQUAL, this.input[this.pos] + this.input[this.readPos])

                } else {
                    tok = this.newToken(TokenType.GT, this.char)
                }
                break
            case '!':
                if (this.peekChar() == "=") {
                    tok = this.newToken(TokenType.NEQUAL, this.input[this.pos] + this.input[this.readPos])
                } else {
                    tok = this.newToken(TokenType.BANG, this.char)
                }
                break
            case '/':
                tok = this.newToken(TokenType.SLASH, this.char)
                break
            case '\0':
                tok = this.newToken(TokenType.EOF, "")
                break
            default:
                if (this.isLetter(this.char.charCodeAt(0))) {
                    const literal = this.readIdentifier()
                    const type = readIDENT(literal)
                    tok = this.newToken(type, literal)
                    return tok
                } else if (this.isDigit(this.char.charCodeAt(0))) {
                    const literal = this.readNumber()
                    const type = identNumber(literal)
                    tok = this.newToken(type, literal)
                    return tok
                }
                tok = this.newToken(TokenType.ILLEGAL, "")
                break
        }
        this.readChar()
        return tok

    }

    private newToken(TokenType: TokenItem, char: string): Token {
        return {
            type: TokenType,
            literal: char
        } as Token
    }
    private readIdentifier(): string {
        let position = this.pos;
        while (this.isLetter(this.char.charCodeAt(0))) {
            this.readChar()
        }

        return this.input.slice(position, this.pos)
    }

    private readNumber(): string {
        let position = this.pos;
        while (this.isDigit(this.char.charCodeAt(0))) {
            this.readChar()
        }

        return this.input.slice(position, this.pos)
    }


    private peekChar(): string {
        if (this.readPos >= this.input.length) {
            return '\0';
        }
        return this.input[this.readPos]
    }


    private isLetter(char: number): boolean {
        return char >= A && char <= Z || char >= a && char <= z || char === _
    }

    private skipWhiteSpace() {
        while (this.char == ' ' || this.char == '\t' || this.char == '\n' || this.char == '\r') {
            this.readChar()
        }
    }

    private isDigit(char: number): boolean {
        return char >= _0 && char <= _9 || char === dot
    }

}

File: RustLan/src/lexer.rs
use crate::token::{Token, TokenType};

struct Lexer<'a> {
    input: &'a [u8],
    pos: usize,
    read_pos: usize,
    ch: u8,
}

impl<'a> Lexer<'a> {
    pub fn new_lexer(input: &'a [u8]) -> Self {
        let mut l = Lexer {
            input,
            pos: 0,
            read_pos: 0,
            ch: b'\0',
        };
        l.read_char();
        return l;
    }
    pub fn new_token(self: &Self, token_type: TokenType) -> Token<'a> {
        Token {
            token_type,
            literal: &self.input[self.pos..self.read_pos],
        }
    }

    pub fn next_token(self: &mut Self) -> Token<'a> {
        let t = match self.ch {
            b'=' => self.new_token(TokenType::ASSIGN),
            b'+' => self.new_token(TokenType::PLUS),
            b'(' => self.new_token(TokenType::LPAREN),
            b')' => self.new_token(TokenType::RPAREN),
            b'{' => self.new_token(TokenType::LBRACE),
            b'}' => self.new_token(TokenType::RBRACE),
            b',' => self.new_token(TokenType::COMMA),
            b';' => self.new_token(TokenType::SEMICOLON),
            0 => self.new_token(TokenType::EOF),
            _ => self.new_token(TokenType::ILLEGAL),
        };
        self.read_char();
        return t;
    }

    pub fn read_char(self: &mut Self) {
        if self.read_pos >= self.input.len() {
            self.ch = 0;
        } else {
            self.ch = self.input[self.read_pos];
        }
        self.pos = self.read_pos;
        self.read_pos += 1;
    }
}

#[cfg(test)]
mod test {
    use super::*; // Import Lexer, Token, TokenType from the outer file

    #[test]
    fn test_next_token() {
        let input = b"=+(){},;";

        // Define our expected output
        let tests = vec![
            (TokenType::ASSIGN, b"="),
            (TokenType::PLUS, b"+"),
            (TokenType::LPAREN, b"("),
            (TokenType::RPAREN, b")"),
            (TokenType::LBRACE, b"{"),
            (TokenType::RBRACE, b"}"),
            (TokenType::COMMA, b","),
            (TokenType::SEMICOLON, b";"),
            (TokenType::EOF, &[0u8]),
        ];

        let mut l = Lexer::new_lexer(input);

        for (expected_type, expected_literal) in tests {
            let tok = l.next_token();

            assert_eq!(tok.token_type, expected_type);
            assert_eq!(tok.literal, expected_literal);
        }
    }
}
